harness/determined/profiler.py

```diff
@@ -0,0 +1,517 @@
+import datetime
+import logging
+import queue
+import threading
+import time
+from typing import Optional, List, Dict, Any
+
+import psutil
+import pynvml
+
+
+from determined.common import api
+from determined.common.api import TrialProfilerMetricsBatch
+
+SYSTEM_METRIC_TYPE_ENUM = "PROFILER_METRIC_TYPE_SYSTEM"
+
+LOG_NAMESPACE = "determined-profiler"
+
+
+
+class ProfilerAgent:
+    """
+    Agent that collects metrics and sends them to the master. It has:
+    - a thread to send data to the Master API (sender_thread)
+    - a thread to collect System Metrics and periodically flush to sender_thread (sys_metric_collector_thread)
+    - [UNIMPLEMENTED] something to batch Timings and periodically flush to sender_thread (timings_batcher)
+
+    The ProfilerAgent needs to be created at the beginning of training and it needs
+    to be notified every time the batch_idx increases. When it is created, it launches
+    the sender_thread and the sys_metric_collector_thread. They will be cleaned up
+    either once end_after_batch is finished or after 5 minutes have passed.
+
+    You can also ship Timings through the ProfilerAgent with the record_timing() method. This
+    functionality has not yet been implemented - we need to batch the timings and periodically
+    flush them to the sender_thread.
+
+    Profiling is only active between start_on_batch and end_after_batch. It will also automatically
+    shut down 5 minutes after starting. When profiling is not active, no system metrics are collected
+    and the record_timing function is a no-op.
+
+    Usage:
+    ```
+    profiler_agent = ProfilerAgent(self, trial_id, agent_id, master_url, start_on_batch, end_after_batch)
+
+    for batch_idx, batch in enumerate(batches):
+        profiler_agent.update_batch_idx(batch_idx)
+
+        # NOTE: Timing API has not been fully developed yet
+        forward_pass_timing = Timing("forward_pass")
+        forward_pass_timing.start()
+        # Do forward pass
+        forward_pass_timing.end()
+        profiler_agent.record_timing(forward_pass_timing)
+    ```
+    """
+    def __init__(self, trial_id: int, agent_id: str, master_url: str, start_on_batch: int, end_after_batch: Optional[int] = None):
+        self.current_batch_idx = 0
+        self.agent_id = agent_id
+        self.trial_id = trial_id
+        self.master_url = master_url
+        self.start_on_batch = start_on_batch
+        self.end_after_batch = end_after_batch
+
+        # Set up timer to stop collecting after 5 minutes
+        self.has_started = False
+        self.has_finished = False
+        self.max_collection_seconds = 300
+        self.shutdown_timer = threading.Timer(self.max_collection_seconds, self._end_collection)
```

Comment:
Kicking off a thread just for this seems like overkill when you could just start the SysMetricCollectorThread with a parameter of 300, so that it just dies if it collects that far.

That would have the extra benefit that you wouldn't the shutdown_lock since `_end_collection` would only get called from the main thread.

Reply:
Collecting Timings and batching them will probably require a second thread to run in parallel to the SysMetricsCollector thread that should also shut down after 5 minutes. In that case it is cleaner to have a single shutdown timer. If not, the cost of having the shutdown timer really isn't that high.

Reply:
Sure, but why would you not just have the collector thread always send a Shutdown message to the sender thread?  That way you really do only have one timer in the system (I'm counting the collector thread as a form of a timer during its normal operation).  You also would get that totally for free.

I think what I am more concerned about here is that there's no way to close down this Timer thread.  It's not that it's *that* expensive (though it is creating an entire thread where a simple `if` statement would suffice); it's not that it's *that* risky to let it just sit around for 5 minutes in the background.  But to me it's definitely an "ease of reasoning about" issue.

If there is one pattern in concurrency that makes concurrency easy to reason about, it's [structured concurrency](http://libdill.org/structured-concurrency.html), which is basically the rule that anything that spins off subroutines will always clean up the subroutines before exiting.  It's easy to reason about because when the parent object is closed, you know that all child objects (threads, subprocesses, whatever) are also closed.  Also, it is easy to test because if you ever run two similar tests in a row you don't have leftovers from one test stepping on the next one.

We didn't used to practice structured concurrency in the harness, but we would run into weird bugs where e.g. unit tests would OOM a GPU if you tried to run the whole unit test suite all at once (https://determinedai.atlassian.net/browse/DET-2026) or there was indeterministic shutdown order between when ZMQ sockets got closed vs when the interpreter shut down, which left cryptic error messages everywhere.

Reply:
You have the Profiler wrapper, then you have two child threads the SysMetricCollector and the TimingsBatcher, both of which depend on a third child thread, SenderThread. Putting the SysMetricCollector in charge of controlling timeouts and the SenderThread doesn't make any sense to me - then you need to synchronize across two threads to shut down correctly when the SysMetricCollector and TimingsBatcher threads otherwise wouldn't need to communicate at all. 

If you insist, I will implement a custom Timer class that can be manually cleaned up, but I feel like that is spending time solving a problem that doesn't really occur. Is the specific case you are concerned about is when you write a test that run the harnesses multiple times inside the same process so the timer persists across harness executions?

Reply:
I'll write a custom Timer class. That seems like the solution that will make us both happy

Reply:
We talked in person. Sticking with the Timer thread for now (custom so it can follows structured concurrency)  so we don't need to introduce communication between the SysMetricCollector and TimingsBatcher threads.